virttest/qemu_devices/qcontainer.py

```diff
@@ -1064,12 +1064,14 @@ def pflash_handler(firmware_name, cmd):
                 path = storage.get_image_filename_filesytem(img_params,
                                                             current_data_dir)
                 img_info = json.loads(qemu_storage.QemuImg(img_params,
-                                                           data_dir.DATA_DIR,
+                                                           current_data_dir,
                                                            path).info(True, "json"))
                 img_name = os.path.basename(img_info.get("filename"))
 
                 pflash_vars_name = "%s_%s_VARS.fd" % (self.vmname, img_name)
                 # Force the image file to be saved under "images" directory
+                # FIXME: Put the pflash_var file and the image file in the same
+                # directory if necessary.
```

Comment:
Can this be implemented? For instance:
```diff
                 # FIXME: Put the pflash_var file and the image file in the same
                 # directory if necessary.
-                pflash_vars_path = img_name + "_VARS.fd"
+                pflash_vars_path = os.path.join(current_data_dir, img_name + "_VARS.fd")
                 pflash0, pflash1 = (firmware_name + '_code',
```
In the current implementation I keep getting inconsistent locations for the pflash var file. Is there a reason to postpone implementing this here?

Reply:
> In the current implementation I keep getting inconsistent locations for the pflash var file.

Really? pflash files should all in `data_dir.DATA_DIR/images`.

Some images are named similarly `image_name = images/f31-${vm_arch_name}`, so if you want to save fd files under `current_data_dir`, I think we can easily use
```
                path = storage.get_image_filename_filesytem(img_params,
                                                            current_data_dir)
                pflash_vars_path = path + "_VARS.fd"
```
To avoid fd files save under like `/var/lib/avocado/data/avocado-vt/` but not `/var/lib/avocado/data/avocado-vt/images/`.

Reply:
Either way is good, the current changes however don't achieve this and I end up with
```
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/ovmf/OVMF_CODE.fd \
    -drive if=pflash,format=raw,file=BusinessServer_64.qcow2_VARS.fd \
```
which is not admissible. The fix I suggested above has been tested so far and it works. I can test your fix too if you want to push it. We should drop the `FIXME` and implement this now rather than later as it leads to undesirable behaviors like above.

Reply:
https://github.com/avocado-framework/avocado-vt/blob/6142776eac1ea69f0b2461759fecf4899f91fb1f/virttest/qemu_devices/qcontainer.py#L1072-L1074

Currently, var files are under `DATA_DIR/images`, I am not sure why `-drive if=pflash,format=raw,file=BusinessServer_64.qcow2_VARS.fd \` will happen.

My test result:
```
    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
    -drive if=pflash,format=raw,file=/var/lib/avocado/data/avocado-vt/images/avocado-vt-vm1_rhel830-aarch64-virtio-scsi.qcow2_VARS.fd \
```

Reply:
> Currently, var files are under `DATA_DIR/images`, I am not sure why `-drive if=pflash,format=raw,file=BusinessServer_64.qcow2_VARS.fd \` will happen.

We are using a custom `images_base_dir` like `images_base_dir = /mnt/local/images/c37_mis1/` which doesn't seem to take effect here.

Also there is no denying that the current implementation of `storage.get_image_filename_filesytem` and all related functions is very hard to read and keeps overwriting the path on each occasion. I am not sure which logical path is taken there but I am not sure it is worth investigating it instead of simply making sure to set the path in a simple manner at this stage.

Reply:
```
2021-03-04 07:40:29,291 test             L0326 DEBUG|     image_format = qcow2
2021-03-04 07:40:29,291 test             L0326 DEBUG|     image_name = image1
2021-03-04 07:40:29,292 test             L0326 DEBUG|     images = image1
2021-03-04 07:40:29,292 test             L0326 DEBUG|     images_base_dir = /mnt/local/images/c37_mis1/

    -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2/aarch64/QEMU_EFI-silent-pflash.raw \
    -drive if=pflash,format=raw,file=/var/lib/avocado/data/avocado-vt/images/avocado-vt-vm1_image1.qcow2_VARS.fd \
```

It seems that `BusinessServer_64.qcow2_VARS.fd` is generated by `pflash_vars_name = "%s_%s_VARS.fd" % (self.vmname, img_name)`, but the next line `pflash_vars_path = os.path.join(data_dir.DATA_DIR, "images", pflash_vars_name)` does not work.
Even if `data_dir.DATA_DIR` is empty, the pflash path should also be `images/BusinessServer_64.qcow2_VARS.fd`, so the result is strange.

Reply:
Yes, so just to summarize the desired behavior and if we both agree: the goal is to be able to define `images_base_dir = /some/path` and thus make the code aware that we want to place the VARS file in the images base directory. Do you agree? This doesn't seem to be achievable with the current changes so far.

Reply:
Sure, I agree with you. This item is okay for me, it can even make the code more concise.

I will update it first and then wait for our results before making a decision.